File: python/03_Longest_Substring_Without_Repeating_Characters.py
'''
Given "abcabcbb", the answer is "abc", which the length is 3.

Given "bbbbb", the answer is "b", with the length of 1.

Given "pwwkew", the answer is "wke", with the length of 3. Note that the answer must be a substring, "pwke" is a subsequence and not a substring.
'''

class Solution(object):
    # get longest substring
    def lengthOfLongestSubstring(self, s):
        last = -1
        maximum = 0
        n = len(s)
        sets = {}
        for i in range(n):
            # last < sets[] is to get rid of cases as "abba"
            if s[i] in sets and last < sets[s[i]]:
                last = sets[s[i]]
            sets[s[i]] = i
            maximum = max(i - last, maximum)
        return maximum

File: c++/03_Longest_Substring_Without_Repeating_Characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int last = -1;
        int maximum = 0;
        int n = s.size();
        map<char, int> dict;
        for (int i = 0; i < n; i++) {
            if ((dict.count(s[i]) != 0) && (last < dict[s[i]])) {
                last = dict[s[i]];
            }
            dict[s[i]] = i;
            maximum = max(i - last, maximum);
        }
        
        return maximum;
        
    }
};